### ISSUE TITLE
Passivation of 4K volumes takes 83 sec due to devmapper/udev constrains

### ISSUE THREAD

**Issue by alex-zadara** [alex-zadara/moving_issues_dst#3]
<a href="https://github.com/lstrugano"><img src="https://avatars.githubusercontent.com/u/1285752?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [lstrugano](https://github.com/lstrugano)**
_Wednesday Mar 23, 2016 at 15:13 GMT_
_Originally opened as https://github.com/zadarastorage/Zadara-VC/issues/8020_

----

I created a VPSA model 1600 with 1.6TB of extended cache. I created 4K volumes via zadara_cfg and then I initiated VC failover. The failover failed. I got the following ticket:

```
Ticket #255808: [HIGH] VPSA VPSA_4K_Vol_256_Mirror (vsa-0000002f, vc-0) at zadara-qa5: VAM failed to demote. 

VC HA could not get VAM demoted to Standby state
```
![image](https://cloud.githubusercontent.com/assets/1285752/13989791/91578896-f11a-11e5-9964-744e2c60b935.png)

VC-0 was rebooted

**Comment by alex-zadara** [comment]
<a href="https://github.com/lstrugano"><img src="https://avatars.githubusercontent.com/u/1285752?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lstrugano](https://github.com/lstrugano)**
_Wednesday Mar 23, 2016 at 15:27 GMT_

----

An important note I forgot to mention is that most probably there was a cloud upgrade running in parallel.

Attached are logs from VPSA and CCs:

ftp://130.30.0.71/zadara-qa5/SNs/qa5-sn1-cc/issue_8020_1458746138--*
ftp://130.30.0.71/zadara-qa5/SNs/qa5-sn2-cc/issue_8020_1458746315--*
ftp://130.30.0.71/zadara-qa5/VPSAs/VPSA_4K_Vol_256_Mirror.vsa-0000002f/issue_8020_1458746045--*
ftp://130.30.0.71/zadara-qa5/VPSAs/VPSA_4K_Vol_256_Mirror.vsa-0000002f/issue_8020_1458746796--*

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 15:50 GMT_

----

Passivation of all the volumes took 83 sec:
```
zadara_vac.log.1	136970	Mar 23 16:25:50 vsa-0000002f-vc-0 zadara_vac: [3519] [rpc  ] zrpc_server_sole_msg_run[99]        : ======> Invoking async message [(zrpc_server_sole_invoke_async_func)vac_passivate_start_req(MsgVoid)] <======
zadara_vac.log	29432	Mar 23 16:27:14 vsa-0000002f-vc-0 zadara_vac: [3519] [     ] vac_passivate_start_req[391]        : PassivateVAC[req_id=0x153a378a358]: start completed, took 83972(msec), err=0
```
However, I am not sure why VAM requested for reboot:
```
Mar 23 16:27:19 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[124]   : detaching DB Vpartition. will execute unmount  (Passivating VAC)
Mar 23 16:27:19 vsa-0000002f-vc-0 zadara_vam: [3771] [     ] vc/vam/api/api_impl_vsa.cpp[1728]   : VamBootManager health status is VC_HEALTH_TRANSIENT(1)
Mar 23 16:27:21 vsa-0000002f-vc-0 zadara_vam: [3771] [     ] vc/vam/api/api_impl_vsa.cpp[1728]   : VamBootManager health status is VC_HEALTH_TRANSIENT(1)
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[2262]  : deferred operation 1  did not complete - counter is 1 
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[2272]  : 1 deferred operations did not complete
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[1152]  : waitForDeferredCompletion failed (10001)
```
I see that VAC successfully completed the passivation.

However, @lstrugano, we do not scale for these numbers right now (although @YHershko will concur). I see that you attached also snapshot policies to each volume. Have you waited until first snapshot created? I think btrfs will choke if we will try to create 4K snapshots in one transaction - see https://github.com/zadarastorage/Zadara-VC/issues/7975.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Wednesday Mar 23, 2016 at 16:01 GMT_

----

Right @alex-zadara , until we understand why passivating 4K volumes takes 83 seconds we cannot blindly say that we cannot scale to those numbers.

Assigning to @lvainblat to analyze what is consuming the time (memory operation on 4K objects should not take 83 sec).

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Wednesday Mar 23, 2016 at 16:02 GMT_

----

@lvainblat , once done please assign to @mmoshe to see why VAM requested reboot

**Comment by alex-zadara** [comment]
<a href="https://github.com/lstrugano"><img src="https://avatars.githubusercontent.com/u/1285752?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lstrugano](https://github.com/lstrugano)**
_Wednesday Mar 23, 2016 at 16:12 GMT_

----

Reproduced again on second try of failover

**Comment by alex-zadara** [comment]
<a href="https://github.com/lstrugano"><img src="https://avatars.githubusercontent.com/u/1285752?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lstrugano](https://github.com/lstrugano)**
_Wednesday Mar 23, 2016 at 16:13 GMT_

----

@lvainblat - we have a default snapshot policy. I just used the default volume creation parameter.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 16:18 GMT_

----

Ok, I see that Liran started creating the 4K volumes at 14:54, attaching default snap policies for each volume. Then at 15:00 snapshot manager triggered creation of 1340 snapshots, because these were the volumes that it had so far. Btrfs was able to do it:
```
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.398218] [3519][cre]btrfs [zbtrfs_blk_virt_cre_snap_flush_writes:912] FS[dm-172]: txn[2761] flush[dm-2855] took 0ms, ret=0
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444763] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:318] FS[dm-172]: txn[2761] took 107ms committers=1 open=0ms blocked=65ms read=0KB flushed=13520/13520Kb
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444767] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:323] FS[dm-172]: txn[2761] rdr1:0 cbg:0 rdr2:0
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444769] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:329] FS[dm-172]: txn[2761] wc:0 wpc:0 wdl:0 wew:0 rdl:0
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444771] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:334] -FS[dm-172] txn[2761] ww:0 cs:60 rdi:0 rdr3:0
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444772] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:339] -FS[dm-172] txn[2761] cfr:2 ccr:0(zts:0) pec:0
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444774] [3519][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:347] FS[dm-172] txn[2761] wrw:42 wrs:0 del-refs[fs-avg=1us avg=2us/5172(5172)] page-cache=21012 Kb
Mar 23 15:00:46 vsa-0000002f-vc-0 kernel: [ 1582.444892] [3519][jrn]btrfs [zjournal_commit:226] zjournal /dev/mapper/vpart-16: pool[1]: set max_transid=2761
```
Then the rest of the volumes were created. Then after an hour, the first 1340 volumes did not need a snapshot (no IO). So VAM created snapshots for the new 2755 volumes. Btrfs did it:
```
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.787095] [3523][cre]btrfs [zbtrfs_blk_virt_cre_snap_flush_writes:912] FS[dm-172]: txn[8202] flush[dm-8365] took 0ms, ret=0
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876646] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:318] FS[dm-172]: txn[8202] took 204ms committers=1 open=0ms blocked=127ms read=0KB flushed=27780/27780Kb
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876650] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:323] FS[dm-172]: txn[8202] rdr1:0 cbg:0 rdr2:0
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876652] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:329] FS[dm-172]: txn[8202] wc:0 wpc:0 wdl:0 wew:0 rdl:0
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876654] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:334] -FS[dm-172] txn[8202] ww:0 cs:115 rdi:0 rdr3:0
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876656] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:339] -FS[dm-172] txn[8202] cfr:8 ccr:1(zts:0) pec:1
Mar 23 16:01:15 vsa-0000002f-vc-0 kernel: [ 5211.876658] [3523][tx]btrfs [ZBTRFS_TXN_COMMIT_PHASE_DONE:347] FS[dm-172] txn[8202] wrw:77 wrs:0 del-refs[fs-avg=3us avg=2us/10701(10701)] page-cache=399960 Kb
Mar 23 16:01:16 vsa-0000002f-vc-0 kernel: [ 5212.153065] [3523][jrn]btrfs [zjournal_commit:226] zjournal /dev/mapper/vpart-16: pool[1]: set max_transid=8202
```
So we already have 8K btrfs subvolumes just after one hour.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 16:19 GMT_

----

@lstrugano snapshot policy has nothing to do with volume passivation problem.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 17:05 GMT_

----

The first thing to try is to increase the amount of VAC API server threads. From the logs, the passivation is very sequential with only 4 threads that VAC has.

Further, for this simple setup, each volume passivation does three main operations:
- deregister from SCST; this is a sysfs call
- remove dm-iostat device mapper
- remove dm-btrfs device mapper

The latter two operations do fork+exec to run the "dmsetup" tool and then interact with udev daemon for appropriate udev event (dmsetup also waits for udev events internally as far as I know). There are few optimizations that I can think of, like:
- send a proprietary IOCTL to dm-iostat to disconnect from dm-btrfs (or from dm-crypt if we have one), by stopping all IOs and executing a dm_put_device() on the lower device (dm-btrfs). Somehow need to ensure that after such IOCTL, the device mapper will not be usable until eventually removed
- send a proprietary IOCTL to dm-btrfs to disconnect from zcache data device (dm_put_device) and from btrfs block-virt file (close the file handle). Same here: need to ensure that device-mapper is not usable anymore for anything.
- once executed for all volumes, we can now go ahead and unmount btrfs (nobody is holding it), while removing all device-mappers in the backround and out of the main passivation path. But we need to somehow ensure that activation won't begin until we fully clean up

Correction: VAC has 5 threads, but one thread is the main thread executing the passivation and it submits work to all other threads.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 18:23 GMT_

----

Some results: increasing VAC API thread pool to 65 threads: the first part of passivation (volumes + CGs) completes in 38 seconds:
```
zadara_vac.log.1	168324	Mar 23 19:31:55 vsa-0000002f-vc-0 zadara_vac: [14523] [     ] __vac__passivate_start[563]         : API Server state set: ACTIVE => PASSIVATING
zadara_vac.log	45934	Mar 23 19:32:33 vsa-0000002f-vc-0 zadara_vac: [14460] [     ] vac_passivation_complete_one[113]   : PassivateVAC[req_id=0x153a481a1e2]: msg group completed, err=0, took 37559(msec)
zadara_vac.log	45935	Mar 23 19:32:33 vsa-0000002f-vc-0 zadara_vac: [14473] [     ] vac_passivate_start_req[359]        : PassivateVAC[req_id=0x153a481a1e2]: all virtual volumes passivated, err=0
zadara_vac.log	58271	Mar 23 19:32:33 vsa-0000002f-vc-0 zadara_vac: [14473] [     ] vac_passivate_start_req[391]        : PassivateVAC[req_id=0x153a481a1e2]: start completed, took 37966(msec), err=0
```
However, when activating with this number of threads, some volumes fail to activate:
```
zadara_vac.log.1	99713	Mar 23 19:37:18 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ] vac_virtual_volume_establish_req[171]: VVolume[0xa5]: establish, lc_state=OBJ_UNKNOWN
zadara_vac.log.1	99714	Mar 23 19:37:18 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ] __vac_virtual_volume_assemble[263]  : VVolume[0xa5]: change lc_state: OBJ_UNKNOWN -> OBJ_CREATING
zadara_vac.log.1	99715	Mar 23 19:37:18 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ] vac_virtual_volume_assemble20_dmbtrfs[478]: VVolume[0xa5]: create dm-btrfs(vol-btrfs-0xa5,pool_devt=253:171,file_path=/btrfs/pool-00000001/cg-000000a6/snap-000000a6/volume-000000a5,gran=2048,flags=0x3,tenant=174,num_sectors=1048576000)
zadara_vac.log.1	108233	Mar 23 19:37:24 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ]*vac_virtual_volume_assemble20_dmbtrfs[548]: VVolume[0xa5]: zdmsetup_create(vol-btrfs-0xa5) failed: No such device(19)
zadara_vac.log.1	108522	Mar 23 19:37:24 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ] vac_virtual_volume_assemble30_dmiostat[775]: VVolume[0xa5]: iostat(vol-iostat-0xa5) dm_devpath=/dev/mapper/vol-iostat-0xa5 canon_devpath=/dev/dm-1132
zadara_vac.log.1	108617	Mar 23 19:37:24 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ] vac_virtual_volume_set_perf_threshold[1651]: VVolume[0xa5]: remove metering alert
zadara_vac.log.1	108618	Mar 23 19:37:24 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ]*__vac_virtual_volume_assemble[395]  : VVolume[0xa5]: change lc_state: OBJ_CREATING -> OBJ_PARTIAL
zadara_vac.log.1	108619	Mar 23 19:37:24 vsa-0000002f-vc-0 zadara_vac: [1655] [vv   ]*vac_virtual_volume_establish_req[205]: VVolume[0xa5]: establish done, lc_state=OBJ_PARTIAL
```
The problem seems to be that udev processing is not fast enough. The device-mapper was created with no issues:
```
Mar 23 19:37:18 vsa-0000002f-vc-0 kernel: [11346.295498] [10086]dm_btrfs [dm_btrfs_ctr:297] Set up RW dev[253:389] over pool=253:171/dm-171 file=/btrfs/pool-00000001/cg-000000a6/snap-000000a6/volume-000000a5 gran=2048(1048576b):11 tenant=174
```
But udev event about its creation was logged only after 5 seconds:
```
Mar 23 19:37:23 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-389 ACTION=remove SEQNUM=107164 USEC_INITIALIZED=53707622 MAJOR=253 DEVPATH=/devices/virtual/block/dm-389 DM_COOKIE=6333377 DM_NAME=vol-btrfs-0xa5 DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-vol-btrfs-0xa5 /dev/mapper/vol-btrfs-0xa5 SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=389 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
Mar 23 19:37:23 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-389 ACTION=remove SEQNUM=107166 USEC_INITIALIZED=53707622 MAJOR=253 DEVPATH=/devices/virtual/block/dm-389 DM_NAME=vol-btrfs-0xa5 DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-vol-btrfs-0xa5 /dev/mapper/vol-btrfs-0xa5 SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=389 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
```
Most probably what happened is that libdevmapper gave up waiting for udev, because I see somebody removing the device:
```
Mar 23 19:37:23 vsa-0000002f-vc-0 kernel: [11351.537221] [26595]dm_btrfs [dm_btrfs_dev_release:411] dev[253:389]: kobject released
Mar 23 19:37:23 vsa-0000002f-vc-0 kernel: [11351.537225] [26595]dm_btrfs [dm_btrfs_dtr:57] dev[253:389]: wait for kobject to be release
Mar 23 19:37:23 vsa-0000002f-vc-0 kernel: [11351.537227] [26595]dm_btrfs [dm_btrfs_dtr:59] dev[253:389]: kobject released
Mar 23 19:37:23 vsa-0000002f-vc-0 kernel: [11351.537231] [26595]dm_btrfs [dm_btrfs_dtr:86] Release dev[253:389] from [253:171]
```
but in VAC we don't do such thing, so must be libdevmapper giving up.

Bottom line is that with large number of threads we overwhelm udev with the amount of events it needs to processs.

Trying 129 threads: passivation fails. Same problem, udev is not fast enough to deliver the deletion events:
```
zadara_vac.log	15954	Mar 23 20:01:24 vsa-0000002f-vc-0 zadara_vac: [1662] [     ] __vac__passivate_start[563]         : API Server state set: ACTIVE => PASSIVATING
zadara_vac.log	94678	Mar 23 20:02:06 vsa-0000002f-vc-0 zadara_vac: [1656] [     ]*vac_passivate_start_req[391]        : PassivateVAC[req_id=0x153a48d440c]: start completed, took 42286(msec), err=17
```
Passivation takes 42 seconds, but fails.

Example:
```
zadara_vac.log	74144	Mar 23 20:01:59 vsa-0000002f-vc-0 zadara_vac: [1654] [vv   ] vac_virtual_volume_passivate_req[331]: VVolume[0xe2d]: passivate, lc_state=OBJ_NORMAL
zadara_vac.log	74145	Mar 23 20:01:59 vsa-0000002f-vc-0 zadara_vac: [1654] [vv   ] vac_virtual_volume_set_perf_threshold[1651]: VVolume[0xe2d]: remove metering alert
                        Mar 23 20:02:05 vsa-0000002f-vc-0 zadara_vac: [1654] [dvw  ]*zdevwait_wait_ex[141]               : Waiter[0x7fa0982fac20], dev_path=[/dev/dm-6906], out_dev_path=[(null)], seqnum=[0], action=[remove] with 2 props FREE(TIMEOUT), 112 waiters remain
zadara_vac.log	81440	Mar 23 20:02:05 vsa-0000002f-vc-0 zadara_vac: [1654] [vv   ]*vac_virtual_volume_teardown20_dmbtrfs[966]: VVolume[0xe2d]: zdmsetup_remove(vol-btrfs-0xe2d) failed: File exists(17)
zadara_vac.log	81441	Mar 23 20:02:05 vsa-0000002f-vc-0 zadara_vac: [1654] [vv   ]*vac_virtual_volume_passivate_req[339]: VVolume[0xe2d]: passivate done, File exists(17)
```
udev event event arrived on time:
```
udev.log	135851	Mar 23 20:02:00 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-6906 ACTION=remove SEQNUM=156053 USEC_INITIALIZED=11029503 MAJOR=253 DEVPATH=/devices/virtual/block/dm-6906 DM_COOKIE=6297971 DM_NAME=vol-btrfs-0xe2d DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-vol-btrfs-0xe2d /dev/mapper/vol-btrfs-0xe2d SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=6906 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
udev.log	135853	Mar 23 20:02:00 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-6906 ACTION=remove SEQNUM=156055 USEC_INITIALIZED=11029503 MAJOR=253 DEVPATH=/devices/virtual/block/dm-6906 DM_NAME=vol-btrfs-0xe2d DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-vol-btrfs-0xe2d /dev/mapper/vol-btrfs-0xe2d SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=6906 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
```
But still VAC did not receive it through udev waiter in the 5-sec timeout that we allocate and gave up.

Conclusion: increasing the number of threads is <b>NOT</b> the way to go IMO.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Mar 23, 2016 at 20:07 GMT_

----

I implemented my proposed fix <b>in a very unsafe way, only for testing</b>. Results are the following:

- Without my fix passivating all volumes takes 60 seconds. Total time of passivation first phase is 63 seconds. In this case additional 3 seconds are spent in stopping iscsi target/NFS/samba

- With my fix passivating all volumes takes 2-3 seconds. Total time of passivation first phase is 15 seconds. 3 seconds are spent in stopping iscsi target/NFS/samba. 10 seconds are spent in passivating zcache (not sure why it took 10 seconds now, but this is unrelated to my change).

Both experiments done with 4 API server threads in VAC.

@YHershko 
Have I answered your question of why volume passivation is slow? (fork-exec + udev + lack of parallellism). As an additional proof: after passivation is completed with my fix, all dm-btrfs and dm-iostat devices are not removed. Removing them with a manual dmsetup command takes:
```
root@vsa-0000002f-vc-0:~# time dmsetup remove_all
real    1m48.451s
user    0m0.000s
sys     0m1.864s
```
more than 100 seconds (probably it is fully sequential)

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Wednesday Mar 23, 2016 at 20:24 GMT_

----

Looks promising! Can this change be made safe?
Let's also have s look at why it takes zcache 10 sec to passivate.
Wow, it will be beautiful if we can take all passivation down to below 10 sec.
Btw can we learn anything from this experiment that will also help in handling large number of BE partitions?

‫

‫ב-23 במרץ 2016, בשעה 22:07, ‏‏alex-zadara ‏<notifications@github.com> כתב/ה:‬

**Comment by alex-zadara** [comment]
<a href="https://github.com/mmoshe"><img src="https://avatars.githubusercontent.com/u/748464?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mmoshe](https://github.com/mmoshe)**
_Thursday Mar 24, 2016 at 08:19 GMT_

----

VAM events in this ticket:
VAM was requested to demote
```
Mar 23 16:25:50 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] vc/vam/api/api_impl_vsa.cpp[1786]   : setVcRole to standby (pre-response)
Mar 23 16:25:50 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] common/mngr/boot_manager.cpp[751]   : bootSwitchRoleAsync called with role standby
```
since passivation took long time vc-ha timed out and requested VAM to go offline before passivation complete.
```
Mar 23 16:26:46 vsa-0000002f-vc-0 vam(VAM:1)[16291]: ERROR: vam-ha(promote): VAM failed to demote within 90 secs (elapsed:90 secs)
Mar 23 16:26:46 vsa-0000002f-vc-0 vam(VAM:1)[16291]: INFO: vam-ha(demote): Resource failed to demote. Triggering Zendesk Ticket & ZSNAP (elapsed:90 secs)

Mar 23 16:26:59 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] common/mngr/boot_manager.cpp[761]   : setting VAM's role to offline will be set to asynchronousy execute ASAP
Mar 23 16:26:59 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] common/mngr/utils/mngr_miscl.cpp[12]: VamDeadline boot_transitions was set (deadlinePeriodMs= 90000 advancementThreasholdMs=24000)
Mar 23 16:26:59 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] common/mngr/boot_manager.cpp[1128]  : about to remove deffered operation type 1 (CallbackId is 18446744073709551615)(MAX_UINT64, i.e., executed not via timer)
Mar 23 16:26:59 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/utils/callback_queue.cpp[279]: entry 18446744073709551615 not found in MngrCallbackQueue generalQueue
Mar 23 16:26:59 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] common/mngr/boot_manager.cpp[2254]  : Start wait for deferred operations to complete. Will time-out in 23 seconds
```
Swiching role to offline waiting for previouse switch (active->standby) to complete. It timed out waiting and tried to passivate VC and set it offline.
```
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[2262]  : deferred operation 1  did not complete - counter is 1 
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[2272]  : 1 deferred operations did not complete
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ]*common/mngr/boot_manager.cpp[1152]  : waitForDeferredCompletion failed (10001)
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [3772] [     ] vc/vam/api/vam_rpc_ha_server.cpp[67]: setVcRoleImplPostResponse of offline done

Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ]*common/mngr/boot_manager.cpp[831]   : offlineActive(not forced) called
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] vc/vam/vam_boot_manager.cpp[433]    : Block all applications & messaging
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] vc/vam/vam_boot_manager.cpp[444]    : Quiesce all applications & messaging
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [exe  ] common/tools/zservice.c[17]         : stop(zadara-eventlog) succeeded: No such file or directory(2)
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [exe  ] common/tools/zservice.c[17]         : STDERR:
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [exe  ] common/tools/zservice.c[17]         :   stop: Unknown instance: 
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ]*common/mngr/boot_manager.cpp[845]   : BootManager is not connectedToDb. No volume passivation done
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] common/mngr/boot_manager.cpp[34]    : HealthVerification start 1 at line 995
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] common/mngr/boot_manager.cpp[1017]  : passivateActive: blocking VamVacInteractionManager, MngrGeneralQueueInterface
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] common/mngr/boot_manager.cpp[1021]  : passivateActive: cleanup of VamVacInteractionManager, MngrGeneralQueueInterface
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] common/mngr/boot_manager.cpp[1028]  : passivateActive: blocking VacSetupStorageInterface
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ] common/mngr/boot_manager.cpp[1030]  : passivateActive: blocking VacInterface
```
Failed to get unblocked VAC interface because it was allready blocked in passivation and set MUST_REBOOT
```
Mar 23 16:27:22 vsa-0000002f-vc-0 zadara_vam: [12969] [     ]*common/mngr/boot_manager.cpp[1033]  : Failed to get the unblocked VacInterface singleton m_state =11, m_curHealthState=0
```
Original passivating thread continue to passivate untill it detected "offline" request
```
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[1066]  : passivateActive: cleanup of VacSetupStorageInterface
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[1068]  : passivateActive: deleting the  pKeptPreviousSinglton
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[1075]  : passivateActive: calling all the unblockForUse
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ]*common/mngr/boot_manager.cpp[1092]  : passivateActive: expecting reboot, no calling of unblockForUse or reinstalling callbacks
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[1096]  : passivateActive: setting the VAM to VC_ROLE_STANDBY
Mar 23 16:27:24 vsa-0000002f-vc-0 zadara_vam: [29150] [     ] common/mngr/boot_manager.cpp[2108]  : bootPassive:621: current boot action 3 is not as expected (1). stopping
```

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Thursday Mar 24, 2016 at 10:38 GMT_

----

Looking again at the 10sec delay, I don't think this is zcache fault. There is again some slowness in libdevmapper/udev/fork/exec.

VAC unmounts btrfs and removes the zcache metadata device, which completes after 10 seconds:
```
Mar 24 12:26:13 vsa-0000002f-vc-0 zadara_vac: [3628] [pool ] vac_storage_pool_fs_teardown[865]   : Pool[0x1]: change fs_state: MOUNTED-OK -> NOT-MOUNTED
Mar 24 12:26:22 vsa-0000002f-vc-0 zadara_vac: [3628] [exe dm] zdmsetup_issue_remove[1865]        : dmsetup remove(pool-zcache-mdata-0x1) succeeded
```
Kernel removes the cache metadata device immediately:
```
Mar 24 12:26:13 vsa-0000002f-vc-0 kernel: [  287.370115] [13628]dm_zcache [pool_dev_dtr:224] Pool[1] - Removing metadata dm[253:172] pool-dev from 253:170
Mar 24 12:26:13 vsa-0000002f-vc-0 kernel: [  287.370127] [13628]dm_zcache [pool_dev_dtr:244] Pool[1] - pool-dev[253:172] removed
```
But udev event about its removal is logged only after 9 seconds:
```
Mar 24 12:26:22 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-172 ACTION=remove SEQNUM=36755 USEC_INITIALIZED=988036 MAJOR=253 DEVPATH=/devices/virtual/block/dm-172 DM_COOKIE=6343921 DM_NAME=pool-zcache-mdata-0x1 DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-pool-zcache-mdata-0x1 /dev/mapper/pool-zcache-mdata-0x1 SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=172 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
Mar 24 12:26:22 vsa-0000002f-vc-0 udev: DEVNAME=/dev/dm-172 ACTION=remove SEQNUM=36757 USEC_INITIALIZED=988036 MAJOR=253 DEVPATH=/devices/virtual/block/dm-172 DM_NAME=pool-zcache-mdata-0x1 DM_UDEV_PRIMARY_SOURCE_FLAG=1 DEVLINKS=/dev/disk/by-id/dm-name-pool-zcache-mdata-0x1 /dev/mapper/pool-zcache-mdata-0x1 SUBSYSTEM=block DM_UDEV_RULES=1 DM_SUSPENDED=0 MINOR=172 DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1 PWD=/ DEVTYPE=disk
```
That's when VAC frees up and proceeds to removing the zcache data device.

So the above approach or similar should be applied also to dm-zcache: disconnect from the pool device and from all devgroups via a proprietary IOCTL.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Thursday Mar 24, 2016 at 14:49 GMT_

----

@YHershko 
With this, the analysis is completed. The root cause is that passivation takes more than 90 sec. When this happens, vc-ha gives up and tells VAM to become offline, while VAM is in the middle of passivation. The handling of such state in VAM can be perhaps improved, but in any case there is nothing much to do if vam-ha gives up on passivation, and we must reboot.
The root cause of the slow passivation is also very clear.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Saturday Mar 26, 2016 at 05:29 GMT_

----

Thanks, @alex-zadara  for the thourough analysis.
As we discussed over emails, please work with @lstrugano to messure the failover time of a VPSA with  1K volumes and 120 BE device. Then we'll decide the priority of this issue (which I am tending to defer to 16.10).
Please assign back to me once you have the results

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Sunday Mar 27, 2016 at 09:28 GMT_

----

@lvainblat 
Can you please look at kern.log.1 in u:\qa\zadara-qa5\VPSAs\VPSA_4K_Vol_256_Mirror.vsa-0000002f\mlx4_ib_crash1458840536--zsnap--vsa-0000002f-vc-0--2016-03-24--19-36-54--active.tar.gz.

It looks like VMs are never able to establish an iSCSI session. VM starts to login:
```
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028048] iscsi-scst: [1714] iscsi_session_alloc[110]: Session ffff880084ae1100 created: target ffff88063fd99400, tid 1, sid 0xdf0000013d0200, initiator iqn.1993-08.org.debian:01:1e2a8cc5782a
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028056] iscsi-scst: [1714] log_params[89]: Negotiated parameters: InitialR2T No, ImmediateData Yes, MaxConnections 1, MaxRecvDataSegmentLength 1048576, MaxXmitDataSegmentLength 262144, 
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028058] iscsi-scst: [1714] log_params[93]:     MaxBurstLength 1048576, FirstBurstLength 65536, DefaultTime2Wait 0, DefaultTime2Retain 0, 
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028059] iscsi-scst: [1714] log_params[99]:     MaxOutstandingR2T 1, DataPDUInOrder Yes, DataSequenceInOrder Yes, ErrorRecoveryLevel 0, 
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028062] iscsi-scst: [1714] log_params[107]:     HeaderDigest None, DataDigest None, OFMarker No, IFMarker No, OFMarkInt 2048, IFMarkInt 2048, RDMAExtensions No
Mar 24 16:01:07 vsa-0000002f-vc-0 kernel: [ 2665.028064] iscsi-scst: [1714] iscsi_tgt_params_set[276]: Target parameters set for session df0000013d0200: QueuedCommands 128, Response timeout 90, Nop-In interval 30, Nop-In timeout 30
```
But then after 23 seconds:
```
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.384226] iscsi-scst: [1714] iscsi_conn_alloc[933]: Creating connection ffff8807eb2fb100 for sid 0xdf0000013d0200, cid 0 (initiator iqn.1993-08.org.debian:01:1e2a8cc5782a)
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.384240] iscsi-scst: [1714] __iscsi_state_change[433]: ***ERROR***: Connection ffff8807eb2fb100 with initiator iqn.1993-08.org.debian:01:1e2a8cc5782a unexpectedly closed!
```
and
```
Mar 24 16:01:31 vsa-0000002f-vc-0 kernel: [ 2688.451060] iscsi-scst: [841] conn_free[816]: Freeing conn ffff8807eb2fb100 (sess=ffff880084ae1100, 0xdf0000013d0200 0, initiator iqn.1993-08.org.debian:01:1e2a8cc5782a)
```
I suspect this is because SCST is configured to create a thread-per-SCST device, because the log is flooded with:
```
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.442425] scst: [535] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-169 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.493247] scst: [540] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-169 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.544241] scst: [544] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-169 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.595251] scst: [549] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-169 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.646253] scst: [550] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-170 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.697233] scst: [552] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-170 started
Mar 24 16:01:29 vsa-0000002f-vc-0 kernel: [ 2686.748407] scst: [553] scst_cmd_thread[5044]: Processing thread scst_zdr-2f-170 started
```
and
```
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.433201] scst: [22609] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-127 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.434455] scst: [26203] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-131 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.435017] scst: [26329] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-134 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.435055] scst: [26480] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-137 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.435303] scst: [27302] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-140 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.435333] scst: [28024] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-143 finished
Mar 24 16:01:30 vsa-0000002f-vc-0 kernel: [ 2688.435919] scst: [28743] scst_cmd_thread[5072]: Processing thread scst_zdr-2f-147 finished
```
This may indicate a scalability issue in SCST. Once you are done with analysis, please return the issue back to me.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Monday Mar 28, 2016 at 04:57 GMT_

----

@alex-zadara @skaushik @lvainblat ,  while looking for a robust solution to be implemented in 16.10, can we increase the number of VAC threads to 32 or 64 only during passivation ? 
Does it make senseto increase the number of VAC threads to 8 or 16 in general for other operations?

**Comment by alex-zadara** [comment]
<a href="https://github.com/skaushik"><img src="https://avatars.githubusercontent.com/u/723984?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [skaushik](https://github.com/skaushik)**
_Monday Mar 28, 2016 at 05:20 GMT_

----

@YHershko Alex noted that bumping up number of VAC threads causes few udev events to be lost. but agree that 4-threads is perhaps less

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Monday Mar 28, 2016 at 05:33 GMT_

----

I saw @alex-zadara note, but that was only during SOD.

**Comment by alex-zadara** [comment]
<a href="https://github.com/lvainblat"><img src="https://avatars.githubusercontent.com/u/748470?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lvainblat](https://github.com/lvainblat)**
_Monday Mar 28, 2016 at 15:36 GMT_

----

By default SCST works in thread pool type `SCST_THREADS_POOL_PER_INITIATOR`. In this mode it create one thread per ITL during login. Linux creates ~20 threads/sec. In this env we have ~2000 ITLs, so it requires ~100sec for login, while default login timeout is 15sec

**Comment by alex-zadara** [comment]
<a href="https://github.com/lvainblat"><img src="https://avatars.githubusercontent.com/u/748470?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lvainblat](https://github.com/lvainblat)**
_Tuesday Mar 29, 2016 at 09:28 GMT_

----

Created #8067 to track the login issue.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Saturday Apr 02, 2016 at 05:32 GMT_

----

from Alex (over emails):

Removing and creating device-mappers with fork/exec through the “dmsetup” tool + waiting for udev events does not scale.
 
There are two approaches that I can think of to remedy the situation:
 
1) This approach improves only the passivation part. Activation part will still be slow.
The approach is basically what I have described in the issue, and tested. The idea is that we do not attempt to remove device mappers in the main passivation flow. Instead we send a proprietary IOCTL to each device mapper, telling it to disconnect from any block devices it holds:
- For dm-iostat this is its lower block device.
- For dm-crypt this is its lower block device.
- For dm-btrfs this is the pool zcache data device and the btrfs file
- For dm-zcache this is the pool device and all the devgroup devices (I believe)
 
The IOCTL code does dm_put_device()/filp_close() to disconnect from the device.
 
After all IOCTLs have completed, we can go ahead and unmount the btrfs pool. Device-mapper objects remain not-removed, and will be removed after passivation completes.
 
My understanding is that main purposes of passivation are:
- clean unmount of all XFS shares to avoid (or shorten) log replay on activation
- clean unmount btrfs to avoid journal replay on activation
- clean passivation of zcache (with whatever disk updates it needs to do to become clean)
- clean disassemble all the raid-groups to avoid resync on activation
- clean unmount the setup volume ext4 to avoid potential mysql corruption etc
 
After all of the above has been done, we can tell vc-ha that passivation is completed. Then we start removing all the Device-Mappers we left behind. We must not allow another activation until we have removed all the Device Mappers. If we fail to remove a Device Mapper, we reboot. Another option is, of course, that we always do a reboot of the passive VC at the end of passivation.
 
In my tests, it took 2-3 seconds to passivate 4K volumes (8K device-mappers) with this approach, instead of 60 seconds with the current code. This is with 4 VAC API threads and no encrypted volumes.
 
Additional thing that we need to take care of: after dm-iostat/dm-crypt/dm-btrfs receives a proprietary IOCTL and disconnects from whatever it is holding, it must fail all subsequent IOs/IOCTLs (if any). This requires for any IO/IOCTL to check that the Device-Mapper is still operational.
 
- Pro: relatively simple approach; perhaps doable for 16.05 (not sure)
- Cons:
- only passivation is handled
- need to patch all Device-Mappers that we use, including dm-crypt, dm-linear, dm-stripe etc. We can decide not to do this for linear/stripe, because they are used only for pools, and we usually have one or two pools only.
- need to handle IO-after-disconnect problem in all Device Mappers that we change
 
2) This approach handles both activation and passivation, i.e., both creation and deletion of Device Mappers.
Instead of using the “dmsetup” tool (with fork-exec) + waiting for udev events to operate the device mappers, we do it ourselves. We implement the same IOCTLs that libdevmapper uses to operate device mappers. They are sent to Device-Mapper control device. Mimicking them does not seem complicated, looking at the kernel side. Note that we cannot use libdevmapper directly, because it also waits for udev events. Instead of waiting for udev events, we also create and remove the device files ourselves with mknod() and unlink() syscalls.
 
For example:
- We send a standard libdevmapper IOCTL to Device-Mapper control device.
- Kernel creates a Device-Mapper object. This is an in-memory operation and it should not wait for anything.
- In some way (perhaps by changing the standard IOCTLs or sending additional IOCTL) we know what is the new device number.
- We call mknod() to create appropriate device file /dev/XXX (and also symbolic link /dev/mapper/XXX if we want to).
- We can now use the device in VAC.
 
Similar flow is for removal or any other operation on Device Mapper (e.g., suspend/resume/get_table).
 
Pros:
- we avoid fork/exec + udev, and have the fastest path possible
- no changes are needed in the Device Mapper target code
- works for any Device-Mapper operation
- both activation and passivation are improved
Cons:
- some changes might be needed in standard Device-Mapper IOCTLs
- reimplement libdevmapper code in VAC for all Device-Mapper operations
- not doable for 16.05
 
Thanks!

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Saturday Apr 02, 2016 at 05:33 GMT_

----

More from Alex:

Final notes on option-2 – for passivation only:
 
- It is possible to fully disable udev events for a Device Mapper device, by setting the “uevent_suppress” kobject flag.
 
- When removing a Device-Mapper device with the core DM ioctl, there is an inherent delay of 10-20 ms per device in blk_cleanup_queue() function. This is a core block layer function. Additional debugging with system tap suggests that at some point acquiring an IRQ-spinlock is the culprit for the delay. However, the system-tap analysis is not fully conclusive.
 
- We have at least two Device-Mappers per volume (with dm-crypt we have 3), so we have an inherent delay of 20-40 msec per volume. With 4 VAC threads and 1K volumes, passivating them takes 8-10 seconds, which is in line.
 
- Trying 65 threads during volume passivation with the core Device Mapper ioctl actually works very well. Passivating 1K volumes takes 1-2 seconds. With 129 threads, it takes 0-1 seconds.
Conclusions:
 
- At least for the passivation, we can go with the generic option-2.
- We need to add functionality in VAC to control the size of the thread-pool during different stages of passivation. Reason is that for MD-raid-groups we still use udev events (unless we find a way to get rid of them too), and generating large amounts of udev events results in udev wait failures. So we need to increase the amount of threads during volume passivation and decrease the amount during MD passivation.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Monday Apr 04, 2016 at 06:57 GMT_

----

@alex-zadara , let's fix this in 16.10. Too risky for 16.05, but still very important for SOD\EOD of large scale VPSA.

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Wednesday Jun 08, 2016 at 17:23 GMT_

----

@ypresente @YHershko 
Let me know which approach you want to do for this issue. Since this issue tracks only passivation, I think we can go with option-2 for removing the Device Mapper only.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Friday Jun 10, 2016 at 20:06 GMT_

----

I think there is a mixup above, and option-1 is only for passivation and optoin-2 is for both activation and passivation.

@alex-zadara , I can't see any measurements of the time it takes to activate 1K (or 4K) volumes. If option-1 will shorten the activation time from dozens of seconds to few seconds, then I vote going with option-1 to have a proper solution long term for a scalable VPSA.
Do you have rough scoping of each option?

**Comment by alex-zadara** [comment]
<a href="https://github.com/alex-zadara"><img src="https://avatars.githubusercontent.com/u/748488?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alex-zadara](https://github.com/alex-zadara)**
_Monday Jun 20, 2016 at 08:26 GMT_

----

@YHershko 
Option-1 addresses only the passivation. Option-2 can address both activation and passivation, or only one of them. Since this particular issue tracks only passivation, I suggested to use option-2 to address only the passivation, which seems to be easier. I don't have scoping at this point.

**Comment by alex-zadara** [comment]
<a href="https://github.com/YHershko"><img src="https://avatars.githubusercontent.com/u/724306?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [YHershko](https://github.com/YHershko)**
_Monday Jun 20, 2016 at 08:58 GMT_

----

I agree that we should go with option-2, but I'd like to see the implication of applying this also for activation.
